src/lib/address_resolve/AddressResolve.h

```diff
@@ -0,0 +1,178 @@
+    // to be processed, so copying their values (i.e. pointers) is not
+    // allowed.
+    NodeLookupHandleBase(const NodeLookupHandleBase &) = delete;
+    NodeLookupHandleBase & operator=(const NodeLookupHandleBase &) = delete;
+
+    void SetListener(NodeListener * listener) { mListener = listener; }
+    NodeListener * GetListener() { return mListener; }
+
+protected:
+    NodeListener * mListener = nullptr;
+};
+
+/// Represents a request to perform a single node lookup
+/// Contains all the information that should be looked for as well as
+/// extra timeout configurations.
+class NodeLookupRequest
+{
+public:
+    NodeLookupRequest() {}
+    NodeLookupRequest(const PeerId & peerId) : mPeerId(peerId) {}
+
+    NodeLookupRequest(const NodeLookupRequest &) = default;
+    NodeLookupRequest & operator=(const NodeLookupRequest &) = default;
+
+    const PeerId & GetPeerId() const { return mPeerId; }
+    System::Clock::Milliseconds32 GetMinLookupTime() const { return mMinLookupTimeMs; }
+    System::Clock::Milliseconds32 GetMaxLookupTime() const { return mMaxLookupTimeMs; }
+
+    /// The minimum lookup time is how much to wait for additional DNSSD
+    /// queries even if a reply has already been received or to allow for
+    /// additional heuristics regarding node choice to succeed (e.g. for a
+    /// ping6 response to be received, if an implementation goes that route)
+    ///
+    /// Implementations for DNSSD may choose to return responses one by one
+    /// for addresses (e.g. Platform mdns does this at the time this was written)
+    /// or different interfaces will return separate 'done resolving' calls.
+    ///
+    /// If the min lookup time is set to 0, implementations are expected to call
+    /// 'OnNodeAddressResolved' as soon as the first DNSSD response is received.
+    NodeLookupRequest & SetMinLookupTime(System::Clock::Milliseconds32 value)
+    {
+        mMinLookupTimeMs = value;
+        return *this;
+    }
+
+    /// The maximum lookup time is how much to wait until a TIMEOUT error is
+    /// declared.
+    ///
+    /// If a DNSSD response is received before this max timeout, then
+    /// OnNodeAddressResolved will be called on listener objects (immediately)
+    /// if the first DNSSD reply arrives after MinLookupTimeMs has expired.
+    NodeLookupRequest & SetMaxLookupTime(System::Clock::Milliseconds32 value)
+    {
+        mMaxLookupTimeMs = value;
+        return *this;
+    }
+
+private:
+    static constexpr System::Clock::Milliseconds32 kMinLookupTimeMsDefault{ 200 };
+    static constexpr System::Clock::Milliseconds32 kMaxLookupTimeMsDefault{ 3000 };
+
+    PeerId mPeerId;
+    System::Clock::Milliseconds32 mMinLookupTimeMs = kMinLookupTimeMsDefault;
+    System::Clock::Milliseconds32 mMaxLookupTimeMs = kMaxLookupTimeMsDefault;
+};
+
+/// These things are expected to be defined by the implementation header.
+namespace Impl {
+class NodeLookupHandle; // MUST derive from NodeLookupHandleBase
```

Comment:
If the interface is irrelevant, and not used by LookupNode, why must it be anything other than `void*` ? It seems like  `virtual CHIP_ERROR LookupNode(const NodeLookupRequest & request, Impl::NodeLookupHandle & handle)` expects a *by value* handle (it takes a reference to the handle) even though it is used opaquely. Why can't the client keep track of the ownership itself, from a void *?

Reply:
Will add comment: I want callers to provide me the memory for storing things. I.e. I do not have pools and if client wants a pool (i.e. they do not want to have one per CASESession or something like that) go ahead and do this.

I do not want to overallocate a 'pool of lookup handle data'.

Reply:
They can give you the memory by giving you a pointer to memory they own.

Reply:
sure, but then I have to void* and align and make sure the memory buffer is sized sufficiently. All doable but likely uglier to use. I am claiming the benefit of that is not as good as type and memory safety this gives.